### ISSUE TITLE
LimitTo on accountCriteria do not work

### ISSUE THREAD

**Issue by tomekl007** [stormpath/stormpath-sdk-java#425]
given : 
>
int limit = 1;
int offset = n;
client.getAccounts(Accounts.criteria().offsetBy(offset).limitTo(limit));

operation returns more than one elements.
If we have 10 accounts, 
setting offset to 3 and limit to 1, return 10 - 3 = 7 records
setting offset to 8 and limit to 1, return 10 - 8 = 2 records,
setting offset to 0 and limit to 1, return 10 - 0 = 10 records.
Property limit is totally ignored

**Comment by lhazlewood** [comment]
What version of the SDK are you using?

**Comment by tomekl007** [comment]
I am using most recent version, that is 1.0.RC7.6

**Comment by tomekl007** [comment]
I've menage to replicate that bug, see my pull request : 
https://github.com/stormpath/stormpath-sdk-java/pull/426

**Comment by lhazlewood** [comment]
Thanks so much for the PR!

**Comment by tomekl007** [comment]
Hi, is there a possibility that this issue will be resolved soon?

**Comment by mrioan** [comment]
Hi @tomekl007 ,

What you are experiencing is not a bug. It is the actual expected behavior. Please take a look at our Javadoc: https://github.com/stormpath/stormpath-sdk-java/blob/master/api/src/main/java/com/stormpath/sdk/query/Criteria.java#L129

The total size of the list is given by the quantity of items matching the criteria like `Accounts.where(Accounts.name().containsIgnoreCase("Foo")`. 

The whole list is divided in pages. `limitTo` sets the quantity of items per page, rather than the size of the whole list. When you iterate over the list, the backend will be contacted every time the page is exhausted and only X quantity of items will be locally available per request (where X is the page size). Paging happens behind the scenes, you are basically given the chance to set the page size so you can tweak the quantity of items that will travel through the wire at once, but besides that, paging is a transparent concept once the list is being traversed.

So, for example, if the query matches 10 elements and each page contains 1 element (as stated by `limitTo(1)`) then you will be contacting the backend 11 times: 1 for each page + 1 to know that the list has finally been completely traversed. However if you specify `limitTo(25)` then the backend will be contacted only once since the underlying page retrieve and contain all the 10 items that the criteria matched.

BTW, you can always invoke `list.getSize()` to know the actual quantity of items matching the criteria that the whole list will contain.

If you agree with what I am explaining here please let us know so we can close this issue.

Thanks!

**Comment by lhazlewood** [comment]
Closing per the 1.0.RC7.7 release.  If you still believe this to be a problem in light of @mrioan's answer, please re-open.

**Comment by rocketraman** [comment]
I just ran into this myself. I understand @mrioan's explanation, but I think this is the wrong behavior and this should be reopened (at the very least the Javadoc needs some additional explanation).

The `CollectionList` returned by the query doesn't have any pagination specific functionality -- its just an `Iterable`. As such, I defaulted to the understanding that the `CollectionList` itself represents the page I requested, and therefore it should stop iterating at the `limitTo` quantity, since that is what I said I wanted (and nothing in the JavaDoc would contradict this understanding). If I wanted a different page represented by `CollectionList`, I would specify different `offset` and `limit` criteria.

As I understand @mrioan's explanation, the CollectionList actually starts at the given `offset` but DOES NOT end at the given `limitTo`. The `limitTo` simply controls an *internal* function of the SDK which determines when the SDK will contact the server for more elements to iterate through, but this is transparent to the code that is doing the iteration.

A simple workaround is to take only the `limitTo` number of elements from the iteration but this is surprising SDK behavior to say the least.
